File: PhotoViewer-ObjC/LandingViewController.m
//
//  landingViewController.m
//  PhotoViewer
//
//  Created by ASIM27 on 1/20/17.
//  Copyright © 2017 km3h. All rights reserved.
//

#import "landingViewController.h"
#import "DataSource.h"

@interface LandingViewController () <UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate>
@property (strong, nonatomic) DataSource *dataSource;
@property (strong, nonatomic) UIPickerView *typePickerView;
@property (strong, nonatomic) UITextField *textFieldSessionType;
@end

@implementation LandingViewController

-(void)loadView
{
    self.view = [[UIView alloc] init];
    
    UILabel *titleLabel = [[UILabel alloc] init];
    titleLabel.text = NSLocalizedString(@"Photo Viewer", nil);
    titleLabel.font = [UIFont boldSystemFontOfSize:30];
    titleLabel.translatesAutoresizingMaskIntoConstraints = NO;
    titleLabel.adjustsFontSizeToFitWidth = YES;
    titleLabel.minimumScaleFactor = 0.5;
    titleLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:titleLabel];
    
    UIButton *button = [[UIButton alloc] init];
    [button setTitle:NSLocalizedString(@"Display Photos", nil) forState:UIControlStateNormal];
    [button addTarget:self action:@selector(displayPhotosButtonTapped:) forControlEvents:UIControlEventTouchUpInside];
    [button setTitleColor:[UIColor brownColor] forState:UIControlStateNormal];
    button.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:button];
    
    [titleLabel.centerYAnchor constraintEqualToAnchor:self.view.centerYAnchor constant:0].active = YES;
    [titleLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant: 20.0].active = YES;
    [titleLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20.0].active = YES;
    
    self.textFieldSessionType =  [[UITextField alloc] init];
    self.textFieldSessionType.borderStyle = UITextBorderStyleLine;
    self.textFieldSessionType.translatesAutoresizingMaskIntoConstraints = NO;
    self.textFieldSessionType.text = @"";
    self.typePickerView = [[UIPickerView alloc] init];
    self.textFieldSessionType.inputView = self.typePickerView;
    self.textFieldSessionType.delegate = self;
    [self.view addSubview:self.textFieldSessionType];
    
    [self.textFieldSessionType.topAnchor constraintEqualToAnchor:titleLabel.bottomAnchor constant:10].active = true;
    [self.textFieldSessionType.centerXAnchor constraintEqualToAnchor:titleLabel.centerXAnchor constant:10].active = true;
    
    [button.topAnchor constraintEqualToAnchor:self.textFieldSessionType.bottomAnchor constant:20].active = true;
    [button.centerXAnchor constraintEqualToAnchor:titleLabel.centerXAnchor constant:0].active = true;
    
    self.dataSource = [[DataSource alloc] init];
    self.dataSource.urlString = @"http://jsonplaceholder.typicode.com/photos";
    self.dataSource.urlSessionType = URLSessionTypeDelegate;
    [self displaySelectedSessionType:self.dataSource.urlSessionType];
    UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(displaySessionTypePicker:)];
    tapRecognizer.numberOfTapsRequired = 1;
    [self.view addGestureRecognizer:tapRecognizer];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.typePickerView.dataSource = self;
    self.typePickerView.delegate = self;
}

-(void)displayPhotosButtonTapped:(id)sender
{
    self.albumsPhotosSplitVC.dataSource = self.dataSource;
    [self presentViewController:self.albumsPhotosSplitVC animated:YES completion:nil];
}

-(void)displaySelectedSessionType:(enum URLSessionType)sessionType
{
    switch (sessionType)
    {
        case URLSessionTypeDelegate:
            self.textFieldSessionType.text = @"Session Type: Delegate";
            break;
        case URLSessionTypeCompletionHandler:
            self.textFieldSessionType.text = @"Session Type: Completion Handler";
            break;
        default:
            break;
    }
}

-(void)displaySessionTypePicker:(id)sender
{
    [self.textFieldSessionType resignFirstResponder];
}

-(NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}

-(NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    return 2;
}

-(NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component
{
    return row == 0 ? @"Delegate" : @"Completion Handler";
}

-(void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    URLSessionType selectedSessionType = row == 0 ? URLSessionTypeDelegate : URLSessionTypeCompletionHandler;
    self.dataSource.urlSessionType = selectedSessionType;
    [self displaySelectedSessionType:selectedSessionType];
}

@end

File: PhotoViewer/LandingViewController.swift
//
//  LandingViewController.swift
//  PhotoViewer
//
//  Created by ASIM27 on 11/16/16.
//  Copyright © 2016 km3h. All rights reserved.
//

import UIKit

class LandingViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource, UITextFieldDelegate {
    
    public var albumsPhotosSplitVC : AlbumsPhotosSplitVC?
    var dataSource : DataSource = DataSource()
    var typePickerView: UIPickerView = UIPickerView()
    var textFieldSessionType : UITextField?
    
    override func loadView()
    {
        self.view = UIView()
        
        let titleLabel = UILabel()
        titleLabel.text = NSLocalizedString("Photo Viewer", comment: "")
        titleLabel.font = UIFont.boldSystemFont(ofSize: 30)
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.adjustsFontSizeToFitWidth = true
        titleLabel.minimumScaleFactor = 0.5
        titleLabel.textAlignment = .center
        self.view.addSubview(titleLabel)
        
        let button = UIButton.init()
        button.setTitle(NSLocalizedString("Display Photos", comment: ""), for: .normal)
        button.addTarget(nil, action: #selector(displayPhotosButtonTapped(sender:)), for: .touchUpInside)
        button.setTitleColor(UIColor.brown, for: .normal)
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        self.view.addSubview(titleLabel)
        button.translatesAutoresizingMaskIntoConstraints = false
        self.view.addSubview(button)
        
        titleLabel.centerYAnchor.constraint(equalTo: self.view.centerYAnchor).isActive = true
        titleLabel.leadingAnchor.constraint(equalTo: self.view.leadingAnchor, constant: 20.0).isActive = true
        titleLabel.trailingAnchor.constraint(equalTo: self.view.trailingAnchor, constant:-20.0).isActive = true
        
        self.textFieldSessionType = UITextField()
        self.textFieldSessionType?.borderStyle = .line
        self.textFieldSessionType?.translatesAutoresizingMaskIntoConstraints = false
        self.textFieldSessionType?.text = ""
        self.textFieldSessionType?.inputView = self.typePickerView
        self.textFieldSessionType?.delegate = self
        self.view.addSubview(self.textFieldSessionType!)
        
        self.textFieldSessionType?.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 10).isActive = true
        self.textFieldSessionType?.centerXAnchor.constraint(equalTo: titleLabel.centerXAnchor).isActive = true
        
        button.topAnchor.constraint(equalTo: (self.textFieldSessionType?.bottomAnchor)!, constant: 20).isActive = true
        button.centerXAnchor.constraint(equalTo: titleLabel.centerXAnchor).isActive = true
        
        dataSource.urlString = "http://jsonplaceholder.typicode.com/photos"
        dataSource.urlSessionType = .Delegate
        self.displaySelectedSessionType(sessionType: dataSource.urlSessionType)
        let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(displaySessionTypePicker(sender:)))
        tapRecognizer.numberOfTapsRequired = 1;
        self.view.addGestureRecognizer(tapRecognizer)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.typePickerView.dataSource = self
        self.typePickerView.delegate = self
    }
    
    func displayPhotosButtonTapped(sender: Any)
    {
        if let controller = self.albumsPhotosSplitVC
        {
            controller.dataSource = self.dataSource
            self.present(controller, animated: true, completion: nil)
        }
    }
    
    func displaySelectedSessionType(sessionType: URLSessionType?)
    {
        switch sessionType
        {
        case .some(.Delegate):
            self.textFieldSessionType?.text = "Session Type: Delegate"
        case .some(.CompletionHandler):
            self.textFieldSessionType?.text = "Session Type: Completion Handler"
        default:
            break
        }
    }
    
    // MARK: - Delegate methods
    func displaySessionTypePicker(sender: Any)
    {
        self.textFieldSessionType?.resignFirstResponder()
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int
    {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int
    {
        return 2
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String?
    {
        return row == 0 ? "Delegate" : "Completion Handler"
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int)
    {
        let selectedSessionType = row == 0 ? URLSessionType.Delegate : URLSessionType.CompletionHandler
        dataSource.urlSessionType = selectedSessionType
        self.displaySelectedSessionType(sessionType: selectedSessionType)
    }    
}
